laplace/baselaplace.py

```diff
@@ -1568,8 +1567,1025 @@ def sample(self, n_samples=100, generator=None):
+            is returned with the mean and the predictive covariance.
+        """
+
+        if self._fitted is False:
+            raise RuntimeError(
+                'Functional Laplace has not been fitted to any '
+                + 'training dataset. Please call .fit method.'
+            )
+
+        if self._recompute_Sigma is True:
+            Warning(
+                'The prior precision has been changed since fit. '
+                + 'Re-compututing its value...'
+            )
+            self._build_Sigma_inv()
+
+        if pred_type not in ['gp']:
+            raise ValueError('Only gp supported as prediction types.')
+
+        if link_approx not in ['mc', 'probit', 'bridge', 'bridge_norm']:
+            raise ValueError(f'Unsupported link approximation {link_approx}.')
+
+        if generator is not None:
+            if (
+                not isinstance(generator, torch.Generator)
+                or generator.device != x.device
+            ):
+                raise ValueError('Invalid random generator (check type and device).')
+
+        # For reward modeling, replace the likelihood to regression and override model state
+        if self.reward_modeling and self.likelihood == 'classification':
+            self.likelihood = 'regression'
+            self.model.output_size = 1
+
+        f_mu, f_var = self._glm_predictive_distribution(
+            x, joint=joint and self.likelihood == 'regression'
+        )
+        # regression
+        if self.likelihood == 'regression':
+            return f_mu, f_var
+        # classification
+        if link_approx == 'mc':
+            return self.predictive_samples(
+                f_mu,
+                f_var,
+                n_samples=n_samples,
+            ).mean(dim=0)
+        elif link_approx == 'probit':
+            kappa = 1 / torch.sqrt(1.0 + np.pi / 8 * f_var.diagonal(dim1=1, dim2=2))
+            return torch.softmax(kappa * f_mu, dim=-1)
+        elif 'bridge' in link_approx:
+            # zero mean correction
+            f_mu -= (
+                f_var.sum(-1)
+                * f_mu.sum(-1).reshape(-1, 1)
+                / f_var.sum(dim=(1, 2)).reshape(-1, 1)
+            )
+            f_var -= torch.einsum(
+                'bi,bj->bij', f_var.sum(-1), f_var.sum(-2)
+            ) / f_var.sum(dim=(1, 2)).reshape(-1, 1, 1)
+            # Laplace Bridge
+            _, K = f_mu.size(0), f_mu.size(-1)
+            f_var_diag = torch.diagonal(f_var, dim1=1, dim2=2)
+            # optional: variance correction
+            if link_approx == 'bridge_norm':
+                f_var_diag_mean = f_var_diag.mean(dim=1)
+                f_var_diag_mean /= torch.as_tensor([K / 2], device=self._device).sqrt()
+                f_mu /= f_var_diag_mean.sqrt().unsqueeze(-1)
+                f_var_diag /= f_var_diag_mean.unsqueeze(-1)
+            sum_exp = torch.exp(-f_mu).sum(dim=1).unsqueeze(-1)
+            alpha = (1 - 2 / K + f_mu.exp() / K**2 * sum_exp) / f_var_diag
```

Comment:
Let's extract these into some util files. They are now duplicates of the ones in `ParametricLaplace`.

Reply:
I noticed that in the ParametricLaplace `__call__` method with MC link, `_glm_predictive_distribution` is called twice: one in `__call__` and another time in `predictive_samples`. In FunctionalLaplace `predictive_samples` is taking directly `f_mu` and `f_var` to avoid this. However, that makes the function difficult to use directly. Should I re-factor the `__call__` function in ParametricLaplace to avoid the double computation and move it to an utils file?

Reply:
Yes, would be great if you could refactor them.